### PR TITLE
fix windows wasm compilation with packages

### PR DESCRIPTION
**Pull request by Chi-EEE**

### CODE REVIEW

**Comment by Chi-EEE** [comment]
cmake.lua error: couldn't find CMAKE_MAKE_PROGRAM
platform.lua error: toolchains was empty with wasm selected

[wasm-test.zip](https://github.com/user-attachments/files/16790392/wasm-test.zip)

File: xmake/modules/package/tools/cmake.lua [hunk 1]
```diff
@@ -1214,8 +1220,9 @@ end
 -- install package
 function install(package, configs, opt)
     opt = opt or {}
-    local cmake_generator = _get_cmake_generator(package, opt)
+    _get_configs_for_generator(package, configs, opt)
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:1223]: we should not call `_get_configs_for_generator `, 

please detect and set ninja generator in _get_cmake_generator, and remove

https://github.com/xmake-io/xmake/blob/2a2ee54fe65cd4035bc25b01391e63170c952c38/xmake/modules/package/tools/cmake.lua#L760-L766

File: xmake/core/platform/platform.lua [hunk 2]
```diff
@@ -151,7 +151,7 @@ end
 -- get the toolchains
 function _instance:toolchains(opt)
     local toolchains = self:_memcache():get("toolchains")
-    if not toolchains then
+    if not toolchains or #toolchains == 0 then
```
Comment (waruqi) [inline xmake/core/platform/platform.lua:154]: why?

File: xmake/modules/package/tools/cmake.lua [hunk 3]
```diff
@@ -601,10 +601,16 @@ function _get_configs_for_wasm(package, configs, opt)
...
-        if mingw_make then
-            table.insert(configs, "-DCMAKE_MAKE_PROGRAM=" .. mingw_make)
+    if is_subhost("windows") then
+        if opt.cmake_generator ~= "Ninja" then
+            local mingw_make = _get_mingw32_make(package)
+            if mingw_make then
+                table.insert(configs, "-DCMAKE_MAKE_PROGRAM=" .. mingw_make)
+            end
+        else
+            local ninja = find_tool("ninja")
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:611]: we can add `_get_ninja(package)` to wrap it.

File: xmake/core/platform/platform.lua (hunk 2 above)
Comment (Chi-EEE) [reply xmake/core/platform/platform.lua:154]: When I selected wasm as the platform, the toolchains list was empty but not nil, which did not allow me to install headeronly packages

File: xmake/modules/package/tools/cmake.lua [hunk 4]
```diff
@@ -762,7 +776,6 @@ function _get_configs_for_generator(package, configs, opt)
         table.insert(configs, "-G")
         if find_tool("ninja") then
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:777]: we need not to continune to find ninja. the cmake generator has been set.

```lua
table.insert(configs, "-G")
table.insert(configs, "MinGW Makefiles")
```

File: xmake/modules/package/tools/cmake.lua [hunk 5]
```diff
@@ -601,10 +608,17 @@ function _get_configs_for_wasm(package, configs, opt)
...
+            local mingw_make = _get_mingw32_make(package)
+            if mingw_make then
+                table.insert(configs, "-DCMAKE_MAKE_PROGRAM=" .. mingw_make)
+            end
+        else
+            local ninja = _get_ninja(package)
+            if ninja then
+                table.insert(configs, "-DCMAKE_MAKE_PROGRAM=" .. ninja)
+            end
         end
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:622]:
```lua
        if opt.cmake_generator == "Ninja" then
            local ninja = _get_ninja(package)
            if ninja then
                table.insert(configs, "-DCMAKE_MAKE_PROGRAM=" .. ninja)
            end
     else
            local mingw_make = _get_mingw32_make(package)
            if mingw_make then
                table.insert(configs, "-DCMAKE_MAKE_PROGRAM=" .. mingw_make)
            end
        end
```

File: xmake/modules/package/tools/cmake.lua [hunk 6]
```diff
@@ -366,6 +366,13 @@ function _get_mingw32_make(package)
     end
 end
 
+-- get ninja
+function _get_ninja(package)
+    local ninja = find_tool("ninja")
+    assert(ninja, "ninja not found!")
+    return ninja.program
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:373]: remove assert

```lua
if ninja then
    return ninja.program
end
```

File: xmake/modules/package/tools/cmake.lua [hunk 7]
```diff
@@ -1142,7 +1149,7 @@ function _get_cmake_generator(package, opt)
         if not cmake_generator then
             if package:has_tool("cc", "clang_cl") or package:has_tool("cxx", "clang_cl") then
                 cmake_generator = "Ninja"
-            elseif is_subhost("windows") and package:is_plat("mingw") then
+            elseif is_subhost("windows") and package:is_plat("mingw", "wasm") then
                 local mingw_make = _get_mingw32_make(package)
                 if not mingw_make and find_tool("ninja") then
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:1156]: _get_ninja

File: xmake/modules/package/tools/cmake.lua [hunk 8]
```diff
@@ -757,13 +771,8 @@ function _get_configs_for_generator(package, configs, opt)
...
         table.insert(configs, _get_cmake_generator_for_msvc(package))
     elseif package:is_plat("wasm") and is_subhost("windows") then
-        -- we attempt to use ninja if it exist
-        -- @see https://github.com/xmake-io/xmake/issues/3771
-        table.insert(configs, "-G")
-        if find_tool("ninja") then
-            table.insert(configs, "Ninja")
-            opt.cmake_generator = "Ninja"
-        else
+        if not find_tool("ninja") then
```
Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:774]: _get_ninja

Comment (waruqi) [inline xmake/modules/package/tools/cmake.lua:774]: we need not find ninja here.

File: xmake/core/platform/platform.lua (hunk 2 above)
Comment (waruqi) [reply xmake/core/platform/platform.lua:154]: it should not be empty.

Comment (waruqi) [reply xmake/core/platform/platform.lua:154]: I can't reproduce this. You can print debug.traceback() when itis empty. 

or revert it and open an issue first. I will merge other patches.

Comment (Chi-EEE) [reply xmake/core/platform/platform.lua:154]: I got this error before adding the line:
```
error: ...amdir\core\sandbox\modules\import\core\tool\compiler.lua:37: cannot get program for cc
stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:973]:
    [...amdir\core\sandbox\modules\import\core\tool\compiler.lua:37]: in function 'load'
    [...amdir\core\sandbox\modules\import\core\tool\compiler.lua:316]:
    [@programdir\modules\package\tools\cmake.lua:886]: in function '_get_envs_for_runtime_flags'
    [@programdir\modules\package\tools\cmake.lua:928]: in function '_get_configs'
    [@programdir\modules\package\tools\cmake.lua:1234]: in function 'install'
    [...\.xmake\repositories\xmake-repo\packages\f\fmt\xmake.lua:70]: in function 'script'
    [...dir\modules\private\action\require\impl\utils\filter.lua:114]: in function 'call'
    [...\modules\private\action\require\impl\actions\install.lua:392]:

  => install fmt 11.0.2 .. failed
error: @programdir\core\main.lua:329: @programdir\modules\async\runjobs.lua:325: ...\modules\private\action\require\impl\actions\install.lua:494: install failed!
stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:973]:
    [...\modules\private\action\require\impl\actions\install.lua:494]: in function 'catch'
    [@programdir\core\sandbox\modules\try.lua:123]: in function 'try'
    [...\modules\private\action\require\impl\actions\install.lua:361]:
    [...modules\private\action\require\impl\install_packages.lua:496]: in function 'jobfunc'
    [@programdir\modules\async\runjobs.lua:241]:

stack traceback:
        [C]: in function 'error'
        @programdir\core\base\os.lua:973: in function 'base/os.raiselevel'
        (...tail calls...)
        @programdir\core\main.lua:329: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:406: in function <@programdir\core\base\scheduler.lua:399>
```

Comment (waruqi) [reply xmake/core/platform/platform.lua]: you can open an issue.

**Pull request merged**